README.md
# README

## Table of contents
* [Dependencies](#dependencies)
* [Getting started](#getting-started)
* [Troubleshooting](#troubleshooting)
* [Task](#task)
* [ToDo](#todo)

## Dependencies
- Python3.4 (or higher)
    - `pytest` library (if you want to run tests)
- Ruby 2.3 or 2.4
- go1.10

## Getting started
0. Your OS:
    - if you are not running on a GNU/Linux OS, you are on your own with debugging

1. Get the application (clone or download)
    - `git clone http://github.com/mrl5/polyglot-task.git`
    - [download](https://github.com/mrl5/polyglot-task/archive/master.zip) and extract
    - go to the application's directory `cd polyglot-task`

2. run `python3 installer.py` (this script will reproduce steps 2, 3, 4)

**OR**

2. Check for dependencies
    - Python: `python --version`
    - Ruby: `ruby --version`
    - Go: `go version`

3. Compile `api.go`
    - make sure that you are inside application's directory `pwd -P | xargs basename` (you should get `polyglot-task
`)
    - `go build api.go`
4. Make sure that `endpoint.rb`, `api` and `worker.py` are executable
    - if you are running on GNU/Linux run `chmod +x endpoint.rb api worker.py`
5. Play around
    - you can start the program by typing `./endpoint.rb` or `ruby endpoint.rb`
6. Logfiles will be stored in folder `polyglot-task-by-JK-aka-mrl5-v20` in your user's home directory (e.g. `/home/kuba/polyglot-task-by-JK-aka-mrl5-v20`)

## Troubleshooting
- `api.go` doesn't compile!
    - download the binary version from [here] (check `go_env.txt` file and check section **4** from "Getting started")
- error: `sh: /home/kuba/test/polyglot-task/api: No such file or directory`
    - there is no `api` binary file in project's directory: check sections **2** and **3** for the solution
- error: `sh: /home/kuba/test/polyglot-task/api: Permission denied`
    - your user has no permission to run `api` file: check section **4** for a solution
    - if you can't change file permission you should clone or extract the project inside your user's home directory
- "I provide valid RPN expression but I get `error` output"
    - if your RPN expression is valid your user has no permission to run `worker.py` file: check section **4** for a solution
    - if you can't change file permission you should clone or extract the project inside your user's home directory
    - if none of above works, try running `python worker.py "<your RPN_expression>"` (e.g. `python worker.py "1 1 +"`) python interpreter should give you more information about what went wrong

## Task
#### Input:

```
2
3 4 +
5 1 2 + 4 * + 3 -
```
Where:
- N - number of input expressions to solve
- f(x1) - first RPN expression to evaluate
- f(x2) - second RPN expression to evaluate
- f(x3) - nth RPN expression to evaluate

#### Output
```
7, 0.235
14, 0.280
```
Where:
- first field: evaluation result
- second field: evaluation time

#### Architecture
1. Endpoint interface for gathering input data (stdin) and output results (stdout)
2. API for getting requests, sending to the worker and logging input and timings. Logs should be
saved to a file.
3. Worker responsible for the process of calculation

## ToDo
#### api.go
- [ ] log StdErr in error.log

[here]: https://drive.google.com/drive/folders/1nweyNIvOPzCxzVGGL3a9n_3ExE0sKnYQ?usp=sharing


CHANGELOG.md
# Changelog
All **notable** changes to this project will be documented in this file.

## [v2.0] - 2018-10-07
- `endpoint.rb` calls API **once**
- `api.go` uses **goroutines** to call `worker.py`
- `installer.py` checks for dependencies and compiles `api.go`
- new `input.log` format: `<request UUID>	<execution time>	<input expression>`
- new logfile: `requests.log`

##  v1.0 - 2018-10-02
- first official release


[v2.0]: https://github.com/mrl5/polyglot-task/compare/v1.0...v2.0


test/test_worker.py
# -*- coding: utf-8 -*-

import pytest
import subprocess
import os
from worker import Worker

__author__ = "mrl5"

"""
Scenario:
    - 'evaluate()' method should return expected values
    - 'evaluate()' method should ignore extra whitespaces
    - worker.py accepts only one argument else raises "Exception" exception
    - 'evaluate()' method should rise "CorruptedRPNExpressionError" custom exception when there are too many operators in RPN expression
    - 'evaluate()' method should rise "CorruptedRPNExpressionError" custom exception if list is not empty after evaluation
    - 'evaluate()' method should rise "CorruptedRPNExpressionError" custom exception if expression is empty
    - 'evaluate()' method should rise "ValueError" when expression has strings or other chars than defined in 'ops' dictionary
    - test if 'evaluate()' method is run from main()
"""


@pytest.fixture(scope="function")
def expected_results():
    dictionary = {
        "3 4 +": 7,
        "5 1 2 + 4 * + 3 -": 14,
        "1 2 +": 3,
        "990 1 2 + *": 2970,
        "1000 990 1 2 + * +": 3970
    }
    return dictionary


@pytest.fixture(scope="function")
def worker_instance():
    worker = Worker()
    return worker

def test_evaluate_method(expected_results, worker_instance):
    real_results = {}
    # create dictionary with real results
    for key, value in expected_results.items():
        result = {key: worker_instance.evaluate(key)}
        real_results.update(result)
    assert expected_results == real_results


def test_additional_whitespaces_in_expression(expected_results, worker_instance):
    # add extra expressions to the expected_results dict
    expected_results.update({"3 4 + ": 7})
    expected_results.update({"3  4 +": 7})
    expected_results.update({"3 4  +": 7})
    expected_results.update({" 3 4  +": 7})
    expected_results.update({"3 4 +\t": 7})
    expected_results.update({"3 \t4 +": 7})
    expected_results.update({"\t3 4 + ": 7})
    expected_results.update({"3\t4\t+": 7})
    real_results = {}
    for key, value in expected_results.items():
        result = {key: worker_instance.evaluate(key)}
        real_results.update(result)
    assert expected_results == real_results


def test_for_IndexError(worker_instance):
    corrupted_expression = "3 4 + -"
    with pytest.raises(worker_instance.CorruptedRPNExpressionError):
        worker_instance.evaluate(corrupted_expression)


def test_too_many_numbers(worker_instance):
    corrupted_expression = "2 3 4 +"
    with pytest.raises(worker_instance.CorruptedRPNExpressionError):
        worker_instance.evaluate(corrupted_expression)


def test_for_empty_input(worker_instance):
    empty_input = " "
    with pytest.raises(worker_instance.CorruptedRPNExpressionError):
        worker_instance.evaluate(empty_input)


def test_disallowed_chars_in_expression(worker_instance):
    corrupted_expression = "abc 4 +"
    with pytest.raises(worker_instance.CorruptedRPNExpressionError):
        worker_instance.evaluate(corrupted_expression)


def test_for_main(worker_instance):
    test_file_location = os.path.dirname(os.path.realpath(__file__))
    tested_file = os.path.join(os.path.dirname(test_file_location), "worker.py")
    expression = "1 1 +"
    cp = subprocess.run(["python3", tested_file, expression], stdout=subprocess.PIPE)
    stdout = bytes(cp.stdout).decode().strip()
    assert stdout == str(worker_instance.evaluate(expression))

test/test_installer.py
# -*- coding: utf-8 -*-

import pytest
import shutil
import os
from installer import Installer

__author__ = "mrl5"

"""
Scenario:
    - '_verify_python()' should return true if version 3
    - '_verify_python()' should return false if other version
    - '_verify_ruby()' should return true if version 2.4
    - '_verify_ruby()' should return true if version 2.3
    - '_verify_ruby()' should return false if other version
    - '_verify_go()' should return true if version 1.10
    - '_verify_go()' should return false if other version 
    - '_compile_api()' should compile 'api.go'
    - binary file created by 'compile_api()' should be in project's directory
    - '_set_executable_permission()' should set executable permissions to the file
"""


@pytest.fixture(scope="function")
def installer_instance():
    installer = Installer()
    return installer


def test_verify_python_success(installer_instance):
    python_version_major = installer_instance._dependencies["python"]["version"]
    assert installer_instance._verify_python(python_version_major) is True


def test_verify_python_fail(installer_instance):
    python_version_major = 2
    assert installer_instance._verify_python(python_version_major) is False


def test_verify_ruby_success(installer_instance):
    ruby_version = installer_instance._dependencies["ruby"]["version"][0]
    assert installer_instance._verify_ruby(ruby_version) is True


def test_verify_ruby_fail(installer_instance):
    ruby_version = "2.5"
    assert installer_instance._verify_ruby(ruby_version) is False


def test_verify_go_success(installer_instance):
    go_version = installer_instance._dependencies["go"]["version"]
    assert installer_instance._verify_go(go_version) is True


def test_verify_go_fail(installer_instance):
    go_version = "1.9"
    assert installer_instance._verify_go(go_version) is False


def test_if_versions_are_stored(installer_instance):
    python_version_major = 2
    ruby_version = "2.5"
    go_version = "1.9"
    installer_instance._verify_python(python_version_major)
    installer_instance._verify_ruby(ruby_version)
    installer_instance._verify_go(go_version)
    test_dict = {
        "python": python_version_major,
        "ruby": ruby_version,
        "go": go_version
    }
    expected_dict_slice = {}
    for k, v in installer_instance._dependencies.items():
        expected_dict_slice.update({k: v["present"]})
    assert test_dict == expected_dict_slice


def test_build_api(tmpdir, installer_instance):
    test_dir = tmpdir.mkdir("test_dir")
    main_dir = os.path.dirname(os.getcwd())
    binary_name = os.path.splitext(installer_instance._api_source)[0]
    api_src = os.path.join(main_dir, installer_instance._api_source)
    api_dest = os.path.join(str(test_dir), installer_instance._api_source)
    shutil.copyfile(api_src, api_dest)
    os.chdir(str(test_dir))
    installer_instance._build_api(api_dest)
    assert os.path.isfile(os.path.join(str(test_dir), binary_name))


def test_set_permission(tmpdir, installer_instance):
    test_file = tmpdir.mkdir("test_dir").join("hello.txt")
    test_file.write("content")
    print(str(test_file))
    installer_instance._set_executable_permission(str(test_file))
    assert os.access(str(test_file), os.X_OK)


api.go
package main

import (
	"fmt"
	"os"
	"errors"
	"path/filepath"
	"os/user"
	"path"
	"os/exec"
	"time"
	"strconv"
	"strings"
	"flag"
	"sync"
	"sort"
)

/* worker related constants */
const (
	WORKER = "worker.py"
	ERROR_MSG = "error"
)

/* logging related constants */
const (
	APP_DIR = "polyglot-task-by-JK-aka-mrl5-v20"
	INPUT_LOG_FILE = "input.log"
	REQUESTS_LOG_FILE = "requests.log"
	APPDIR_PERMISSION = 0750   //owner can do anything with directory, group member can list directory and see rights, others wont see dir content
	LOGFILES_PERMISSION = 0640 //RW for owner, R for group, nothing for others (0640)
)

var endpointUUID string
var logsDir = ""
var pathToInputLog string
var pathToRequestsLog string

func setFlags() {
	/* flags related constants */
	const (
		defaultUUID = "internal"
		uuidUsage = "Universally Unique IDentifier of the endpoint request"
	)

	/* UUID */
	flag.StringVar(&endpointUUID, "uuid", defaultUUID, uuidUsage)
	flag.StringVar(&endpointUUID, "u", defaultUUID, uuidUsage+" (shorthand)")

	/* parse all defined flags */
	flag.Parse()
}

func checkError(e error) {
	if e != nil {
		panic(e)
	}
}

func checkEnvironment(usrDir string) {
	if logsDir == "" {
		logsDir = path.Join(usrDir, APP_DIR)
	}
	pathToInputLog = path.Join(logsDir, INPUT_LOG_FILE)
	pathToRequestsLog = path.Join(logsDir, REQUESTS_LOG_FILE)

	/* create logs dir if doesn't exist, ignore error when it exists */
	_ = os.Mkdir(logsDir, APPDIR_PERMISSION)

	/* check if log files exist, create when they don't */
	os.OpenFile(pathToInputLog, os.O_RDONLY|os.O_CREATE, LOGFILES_PERMISSION)
	os.OpenFile(pathToRequestsLog, os.O_RDONLY|os.O_CREATE, LOGFILES_PERMISSION)
}

func checkInput(noOfArgs int) (int, error) {
	switch noOfArgs {
	default:
		return 0, nil
	case 0:
		return -1, errors.New("API error: expected at least one argument")
	}
}

func logInput(input string, elapsedProcessTime string) {
	f, err := os.OpenFile(pathToInputLog, os.O_APPEND|os.O_WRONLY, LOGFILES_PERMISSION)
	checkError(err)
	defer f.Close()

	/* uuid elapsed-time input */
	logLine := endpointUUID + "\t" + elapsedProcessTime + "\t" + input + "\n"
	if _, err = f.WriteString(logLine); err != nil {
		panic(err)
	}
}

func logRequest(requestStart time.Time, noOfArguments int) {
	f, err := os.OpenFile(pathToRequestsLog, os.O_APPEND|os.O_WRONLY, LOGFILES_PERMISSION)
	checkError(err)
	defer f.Close()

	/* date uuid elapsed-time number-of-args */
	logLine := "[" + requestStart.UTC().String() + "]\t" + endpointUUID + "\t"  +
		strconv.FormatFloat(time.Since(requestStart).Seconds(), 'f', 3, 64) + "\t" +
		strconv.Itoa(noOfArguments) + "\n"
	if _, err = f.WriteString(logLine); err != nil {
		panic(err)
	}
}


/* goroutines framework */
type workerReturn struct {
	id int
	err error
	output []byte
	elapsedTime time.Duration
}

type workerReturnArr []workerReturn
var workerReturns workerReturnArr

func (wr workerReturnArr) Len() int {
	return len(wr)
}

func (wr workerReturnArr) Less(i, j int) bool {
	return wr[i].id < wr[j].id
}

func (wr workerReturnArr) Swap(i, j int) {
	wr[i], wr[j] = wr[j], wr[i]
}
/* /goroutines framework */

func executeWorker(pathToWorker string, argument string, id int, wg *sync.WaitGroup) {
	start := time.Now()
	workerProcess := exec.Command(pathToWorker, argument)
	output, err := workerProcess.Output()
	/* add return to the array of returns */
	workerReturns = append(workerReturns, workerReturn{id, err, output, time.Since(start)})
	wg.Done()
}

func main() {
	start := time.Now()
	setFlags()

	/* path to the directory of binary version of this program */
	workDir, workDirErr := filepath.Abs(filepath.Dir(os.Args[0]))
	checkError(workDirErr)
	pathToWorker := path.Join(workDir, WORKER)

	usrEnv, usrEnvErr := user.Current()
	checkError(usrEnvErr)
	checkEnvironment(usrEnv.HomeDir)

	noOfArgs := len(flag.Args())

	if _, argErr := checkInput(noOfArgs); argErr == nil {
		/* sync.WaitGroup waits for a group of goroutines to finish */
		var wg sync.WaitGroup
		wg.Add(noOfArgs)

		for i := 0; i < noOfArgs; i++ {
			go executeWorker(pathToWorker, flag.Arg(i), i, &wg)
		}
		/* wait until all executeWorker processes will end */
		wg.Wait()
		/* FIFO sorting (by id) */
		sort.Sort(workerReturns)

		for i := 0; i < noOfArgs; i++ {
			var result string
			formattedTime := strconv.FormatFloat(
				workerReturns[i].elapsedTime.Seconds(), 'f', 3, 64)

			if workerReturns[i].err != nil {
				result = ERROR_MSG
			} else {
				result = strings.Trim(
					string(workerReturns[i].output), "\n")
			}
			fmt.Println(result + ", " + formattedTime)
			logInput(flag.Arg(i), formattedTime)
		}

	} else {
		fmt.Println(argErr)
	}
	logRequest(start, len(flag.Args()))
}


kata.go
package main

// https://stackoverflow.com/questions/19208725/example-for-sync-waitgroup-correct
// https://stackoverflow.com/questions/3387273/how-to-implement-resizable-arrays-in-go
// https://gist.github.com/dnutiu/a899e48c95ff80fe98bada566e03251e

import (
	"fmt"
	"sync"
	"time"
	"sort"
)

type mytype struct {
	id, param int
	//err error
	//output []byte
	//elapsedTime time.Duration
}

type mytypeArr []mytype
var returns mytypeArr

func dosomething(id int, millisecs int, wg *sync.WaitGroup) {
	duration := time.Duration(millisecs) * time.Millisecond
	fmt.Println("START: Function in background, duration:", duration)
	time.Sleep(duration)
	fmt.Println("DONE: Function in background, duration:", duration)
	returns = append(returns, mytype{id, millisecs})
	wg.Done()
}

func (s mytypeArr) Len() int {
	return len(s)
}

func (s mytypeArr) Less(i, j int) bool {
	return s[i].id < s[j].id
}

func (s mytypeArr) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

func main() {
	var wg sync.WaitGroup
	const delta = 4
	wg.Add(delta)

	params := [delta]int{6000, 1500, 4000, 2000}

	for i := 0; i < delta; i++ {
		go dosomething(i, params[i], &wg)
	}

	wg.Wait()
	fmt.Println("Done")

	fmt.Println(returns)
	fmt.Println(sort.IsSorted(returns))
	sort.Sort(returns)
	fmt.Println(returns)
	fmt.Println(sort.IsSorted(returns))
	sort.Sort(sort.Reverse(returns))
	fmt.Println(returns)

}

worker.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# based on Yehonathan Sharvit blog post: https://blog.klipse.tech/python/2016/09/22/python-reverse-polish-evaluator.html
__author__ = "mrl5"


class Worker:
    def __init__(self):
        # arithmetic operators as lambda functions (anonymous functions)
        self._ops = {
            "+": (lambda a, b: a + b),
            "-": (lambda a, b: a - b),
            "*": (lambda a, b: a * b),
            "/": (lambda a, b: a / b)
        }

    def evaluate(self, expression):
        tokens = expression.split()
        stack = []
        try:
            for token in tokens:
                if token in self._ops:
                    number2 = stack.pop()
                    number1 = stack.pop()
                    result = self._ops[token](number1, number2)
                    stack.append(result)
                else:
                    stack.append(int(token))
            output = stack.pop()
            # stack should be empty at this point
            if stack:
                raise self.CorruptedRPNExpressionError(expression)
        except IndexError as ie:
            raise self.CorruptedRPNExpressionError(expression) from ie
        except ValueError as ve:
            raise self.CorruptedRPNExpressionError(expression) from ve
        return output

    class CorruptedRPNExpressionError(Exception):
        """
        Custom exception when Reverse Polish Notation expression has non-valid syntax
        """
        def __init__(self, expression=None):
            self.strerror = "Expression has non-valid RPN syntax: {}".format(expression)


def main(args):
    allowed_no_of_args = 1
    if len(args) != allowed_no_of_args:
        raise Exception("Script accepts one and only one argument.")
    worker = Worker()
    expression = args[0]
    print(worker.evaluate(expression))


if __name__ == "__main__":
    import sys
    sys.exit(main(sys.argv[1:]))


endpoint.rb
#!/usr/bin/env ruby

require 'securerandom'

# stores path of this script
script_path = File.dirname(File.expand_path $0)
api_name = "api"
api_path = File.join(script_path, api_name)
# global variables
$api_cmd = api_path + " " + "--uuid=" + SecureRandom.uuid

def welcome_message()
	puts "=== Reverse Polish notation app ==="
	puts ""
	puts "Input syntax:"
	puts "\tNumber of RPN expressions"
	puts "\t1st RPN expression"
	puts "\t2nd RPN expression"
	puts "\t..."
	puts "\tNth RPN expression"
	puts ""
	puts "If you change your mind, hit Ctrl+C to end this app"
	puts ""
	puts "Input:"
end

def check_args()
	if ARGV.length > 0
		abort("No arguments allowed. Aborting.")
	end
end

def init_func()
	# save input and (.chomp) remove carriage return from end of the input string
	inp = gets.chomp
	# convert input to integer, if inp is a string result will be 0
	no_of_expressions = inp.to_i
	return no_of_expressions
end

def get_RPN_expressions(no_of_exp)
	expressions = Array.new
	no_of_exp.times do
		expression = "\'" + gets.chomp.gsub(/\t+/, ' ') + "\'"
		expressions.push(expression)
	end
	return expressions
end

def pass_to_api(arguments)
	system( $api_cmd + " " + arguments.join(" ") )
end

# main
check_args
welcome_message
no_of_expressions = init_func
if no_of_expressions.is_a? Integer
	if no_of_expressions > 0
		expressions = get_RPN_expressions(no_of_expressions)
		puts ""
		puts "Output:"
		pass_to_api(expressions)
	else
		raise("Number of RPN expressions must be greather than 0")
	end
else
	raise("First line must be an integer")
end


installer.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from __future__ import print_function

import sys
import subprocess
import re
import os
from time import sleep

__author__ = "mrl5"


class Installer:
    """
    Class for installer
    """

    def __init__(self):
        self._dependencies = {
            "python": {"version": 3, "present": None},
            "ruby": {"version": ["2.3", "2.4"], "present": None},
            "go": {"version": "1.10", "present": None, "status": False}
        }
        self._api_source = "api.go"
        self._work_files = ["worker.py", "endpoint.rb", os.path.splitext(self._api_source)[0]]

    def _verify_python(self, sys_version_major):
        success_msg = "[OK]\tfound Python {}.{}.{}".format(
            sys.version_info.major, sys.version_info.minor, sys.version_info.micro)
        fail_msg = "[Error]\tfound Python {}.{}.{}. Run this script using 'python3' command.".format(
            sys.version_info.major, sys.version_info.minor, sys.version_info.micro)
        self._dependencies["python"]["present"] = sys_version_major
        verification = self._dependencies["python"]["version"] == sys_version_major
        print(success_msg) if verification else print(fail_msg)
        return verification

    def _verify_ruby(self, ruby_version):
        success_msg = "[OK]\tfound Ruby {}".format(ruby_version)
        fail_msg = "[Error]\tfound Ruby {}. Make sure that `ruby` command leads to the {} version".format(
            ruby_version, " or ".join(self._dependencies["ruby"]["version"]))
        self._dependencies["ruby"]["present"] = ruby_version
        verification = ruby_version in self._dependencies["ruby"]["version"]
        if ruby_version:
            print(success_msg) if verification else print(fail_msg)
        return verification

    def _verify_go(self, go_version):
        success_msg = "[OK]\tfound Go {}".format(go_version)
        fail_msg = "[Error]\tfound Go {} ({} required).".format(
            go_version, self._dependencies["go"]["version"])
        self._dependencies["go"]["present"] = go_version
        verification = self._dependencies["go"]["version"] == go_version
        if go_version:
            print(success_msg) if verification else print(fail_msg)
        self._dependencies["go"]["status"] = verification
        return verification

    def _verify_deps(self):
        print("Checking for dependencies ...")
        sleep(0.5)
        p = self._verify_python(sys.version_info.major)
        sleep(0.5)
        r = self._verify_ruby(get_ruby_version())
        sleep(0.5)
        g = self._verify_go(get_go_version())
        return p and r and g

    def _build_api(self, path_to_api_src):
        os.chdir(os.path.dirname(path_to_api_src))
        print("\nBuilding {} ...".format(self._api_source))
        build_cmd = ["go", "build", path_to_api_src]
        go_build = subprocess.run(build_cmd)
        return go_build.returncode

    def _set_executable_permission(self, path_to_file):
        """
        Sets execute permission for the owner and for a group
        """
        octal_permission = 0o750 # rwxr-x---
        try:
            os.chmod(path_to_file, octal_permission)
            print("[OK]\t{}".format(path_to_file))
            ret_code = 0
        except OSError as oe:
            print("[Error]\t{}: {}".format(oe.filename, oe.strerror))
            ret_code = 1
        return ret_code

    def _quit_msg(self, deps_err, perms_err):
        """
        Prints quit message
        :param deps_err: dependencies error (True/False)
        :param perms_err: permissions error (True/False)
        """
        if not(deps_err | perms_err):
            sleep(0.5)
            print("\nSuccess. Application is ready to use.")
        else:
            sleep(0.5)
            sys.stderr.write("\nNot all dependencies were met. Application might not work properly") if deps_err else False
            sys.stderr.write("\nNot all files have execute permissions.") if perms_err else False
            sys.exit(1)

    def install(self):
        """
        Creates environment for the polyglot-task
        """
        this_file_dir = os.path.dirname(os.path.realpath(__file__))
        path_to_api_src = os.path.join(this_file_dir, self._api_source)
        deps_err = not self._verify_deps()
        if self._dependencies["go"]["status"]:
            sleep(0.5)
            go_build_return_code = self._build_api(path_to_api_src)
        else:
            if self._dependencies["go"]["present"]:
                build_anyways = query_yes_no(
                    "Do you want to compile sources using Go {}?".format(self._dependencies["go"]["present"]))
                go_build_return_code = self._build_api(path_to_api_src) if build_anyways else sys.exit("Aborting.")
            else:
                sleep(2)
                sys.exit("Could not call `go` from the console. Aborting.")
        print("Done.") if go_build_return_code == 0 else sys.exit("\nBuilding {} failed. Aborting.".format(self._api_source))

        print("\nSetting execute permissions ...")
        os.chdir(this_file_dir)
        perms_err = 0
        for f in self._work_files:
            sleep(0.5)
            perms_err = perms_err | self._set_executable_permission(f)
        self._quit_msg(deps_err, bool(perms_err))


def get_ruby_version():
    """
    :return: Ruby version
    """
    cmd = "ruby"
    ruby_version = None
    try:
        rv = subprocess.run([cmd, "--version"], stdout=subprocess.PIPE)
        stdout = bytes.decode(rv.stdout).strip()
        if re.match(r'''
                    ^       # start of string
                    \b      # start of whole word
                    ruby    # "ruby" string
                    \b      # end of whole word
                    ''', stdout, re.VERBOSE):
            ruby_version = re.sub(r'''
                                ^       # start of string
                                \b      # start of whole word
                                ruby    # "ruby" string
                                \b      # end of whole word
                                \s      # one whitespace
                                (       # start of grouping
                                [0-9]   # a digit
                                \.      # "." string
                                [0-9]   # a digit
                                )       # end of grouping
                                .*      # any character (except line break) zero or more times
                                ''', "\\1", stdout, 0, re.VERBOSE)
    except FileNotFoundError:
        print("[Error]\t`{}` command not found. Make sure that ruby is installed and added to the PATH".format(cmd))
    return ruby_version


def get_go_version():
    """
    :return: Go version
    """
    cmd = "go"
    go_version = None
    try:
        rg = subprocess.run([cmd, "version"], stdout=subprocess.PIPE)
        stdout = bytes.decode(rg.stdout).strip()
        if re.match(r'''
                    ^           # start of string
                    go          # "go" string
                    \s          # one whitespace
                    version     # "version" string
                    \s          # one whitespace
                    ''', stdout, re.VERBOSE):
            go_version = re.sub(r'''
                                ^           # start of string
                                go          # "go" string
                                \s          # one whitespace
                                version     # "version" string
                                \s          # one whitespace
                                go          # "go" string
                                (           # start of grouping
                                [0-9]       # a digit
                                \.          # "." string
                                [0-9]+      # one or more digits
                                )           # end of grouping
                                .*          # any character (except line break) zero or more times
                                ''', "\\1", stdout, 0, re.VERBOSE)
    except FileNotFoundError:
        print("[Error]\t`{}` command not found. Make sure that go is installed and added to the PATH".format(cmd))
    return go_version


def query_yes_no(question):
    """
    :param question: question to be answered
    :return: answer
    """
    valid = {"yes": True, "y": True,
             "no": False, "n": False}

    while True:
        sys.stdout.write("{}: [yes/no] ".format(question))
        choice = input().lower()
        if choice == '':
            return True
        elif choice in valid:
            return valid[choice]
        else:
            print("Try again.")


def main(args):
    allowed_no_of_args = 0
    sys.exit("Installer is written for Python 3. Aborting.") if sys.version_info.major < 3 else False
    sys.exit("Script doesn't accept any arguments.") if len(args) != allowed_no_of_args else False
    installer = Installer()
    installer.install()


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
